### PR TITLE
refactor backfill triggering

### PR DESCRIPTION
**Pull request by danielnorberg**
# Hey, I just made a Pull Request!

## Description
Refactor backfill trigger manager to:
* Make code easier to understand
* Give more consistent behavior when there are concurrent updates

## Checklist for PR author(s)

- [ ] Changes are covered by unit test
- [ ] All tests pass
- [ ] Code coverage check passes
- [ ] Error handling is tested
- [ ] Errors are handled at the appropriate layer
- [ ] Errors that cannot be handled where they occur are propagated
- [ ] (optional) Changes are covered by system test
- [ ] Relevant documentation updated
- [ ] This PR has NO breaking change to public API
- [ ] This PR has breaking change to public API and it is documented

## Checklist for PR reviewer(s)

- [ ] This PR has been incorporated in release note for the coming version
- [ ] Risky changes introduced by this PR have been all considered

### CODE REVIEW

File: styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java [hunk 1]
```diff
@@ -135,85 +139,95 @@ void tick0() {
   private void triggerAndProgress(Backfill backfill) {
     // Do not include all backfill spans in parent tick span to avoid it growing too big
     tracer.spanBuilderWithExplicitParent("Styx.BackfillTriggerManager.triggerAndProgress", null)
-        .startSpanAndRun(() -> triggerAndProgress0(backfill));
+        .startSpanAndRun(() -> Try.run(() -> triggerAndProgress0(backfill)));
```
Comment (NatashaL) [inline styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java]: It looks to me like Try.run will swallow unchecked exceptions without logging them. How about just using `guard` like above, which will log any uncaught exceptions?

File: styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java [hunk 2]
```diff
@@ -135,85 +139,95 @@ void tick0() {
...
+
+    // Look up the workflow. Halts the backfill if the workflow does not exist.
+    var workflowOpt = readBackfillWorkflowOrHalt(backfill.id());
+    if (workflowOpt .isEmpty()) {
       return;
     }
+    var workflow = workflowOpt.orElseThrow();
 
-    final Workflow workflow = workflowOpt.get();
+    for (int i = 0; i < MAX_INSTANCES_TICK_TRIGGER; i++) {
```
Comment (NatashaL) [inline styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java]: So you're capping the number of instances to run per backfill per tick. Is that to make sure that bigger backfills don't starve others? Any other reason?

**Comment by NatashaL** [comment]
Also, really nice comments describing the code! Makes it much easier to read and understand the reasoning

File: styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java [hunk 3]
```diff
@@ -135,85 +139,95 @@ void tick0() {
...
-    if (!workflowOpt.isPresent()) {
-      LOG.debug("workflow not found for backfill {}, halt it.", backfill);
-      storeBackfill(backfill.builder().halted(true).build());
+    // Check how many instances for this backfill are currently active
+    // Note: getActiveStatesByTriggerId is eventually consistent and may not return all triggered instances.
+    var alreadyActiveInstances = stateManager.getActiveStatesByTriggerId(backfill.id()).size();
+
+    // Look up the workflow. Halts the backfill if the workflow does not exist.
+    var workflowOpt = readBackfillWorkflowOrHalt(backfill.id());
+    if (workflowOpt .isEmpty()) {
```
[the remark this answers is not in the document]
Comment (danielnorberg) [reply styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java]: 😅

File: styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java (hunk 1 above)
Comment (danielnorberg) [reply styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java]: `guard` doesn't allow checked exceptions

File: styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java (hunk 2 above)
Comment (danielnorberg) [reply styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java]: Non-starvation is one reason. Another is that the `alreadyActiveInstances` might get stale.

File: styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java (hunk 1 above)
Comment (danielnorberg) [reply styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java]: Now propagating the exception though

File: styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java (hunk 2 above)
Comment (danielnorberg) [reply styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java]: Added a comment.

**Comment by NatashaL** [comment]
I tried but couldn't spot why the test fails. @danielnorberg please take a look before merging. Other than that, it looks good 👍

**Comment by danielnorberg** [comment]
The test fails because [reading the workflow in the backfill transaction](https://github.com/spotify/styx/pull/744/files#diff-7724660cca477426c555a62afcc34313R185) conflicts with the [instance triggering transaction](https://github.com/spotify/styx/pull/744/files#diff-7724660cca477426c555a62afcc34313R221). 

There is a [TODO about doing triggering in the backfill transaction](https://github.com/spotify/styx/pull/744/files#diff-7724660cca477426c555a62afcc34313R213), but I'm not sure I want to undertake that refactoring in this PR. For now the simplest thing to do is to just revert https://github.com/spotify/styx/pull/744/commits/ec2cd5558f1dfd54e7f58995262012afa7ad696a which moved the workflow reading into the backfill transaction.

**Pull request merged**